src/org/opensolaris/opengrok/authorization/AuthorizationStack.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * See LICENSE.txt included in this distribution for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+ /*
+ * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
+ */
+package org.opensolaris.opengrok.authorization;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.function.Predicate;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.opensolaris.opengrok.configuration.Nameable;
+import org.opensolaris.opengrok.logger.LoggerFactory;
+
+/**
+ * This is a subclass of {@link AuthorizationEntity} implementing the methods to
+ * be able to contain and making decision for:
+ * <ul>
+ * <li>other stacks</li>
+ * <li>plugins</li>
+ * </ul>
+ *
+ * @author Krystof Tulinger
+ */
+public class AuthorizationStack extends AuthorizationEntity {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AuthorizationStack.class);
+
+    private List<AuthorizationEntity> stack = new ArrayList<>();
+
+    public AuthorizationStack() {
+    }
+
+    public AuthorizationStack(AuthControlFlag flag, String name) {
+        this.flag = flag;
+        this.name = name;
+    }
+
+    /**
+     * Get the value of stack
+     *
+     * @return the current stack
+     */
+    public List<AuthorizationEntity> getStack() {
+        return stack;
+    }
+
+    /**
+     * Set the value of stack
+     *
+     * @param s the new stack
+     */
+    public void setStack(List<AuthorizationEntity> s) {
+        this.stack = s;
+    }
+
+    /**
+     * Add a new authorization entity into this stack.
+     *
+     * @param s new entity
+     */
+    public void add(AuthorizationEntity s) {
+        this.stack.add(s);
+    }
+
+    /**
+     * Remove the given authorization entity from this stack.
+     *
+     * @param s the entity to remove
+     */
+    public void remove(AuthorizationEntity s) {
```

Comment:
how does the unload/remove work in general w.r.t. incoming requests ? If the `Configuration` is being reloaded and the authorization stack deconstructed what mechanism is used to ensure that the evaluation of the request waits until the stack is fully loaded ?